File: scala/src/main/scala/leetcode/ArraysAndHashing/TwoSum.scala
package leetcode.ArraysAndHashing

object TwoSum {
  def twoSum(nums: Array[Int], target: Int): Array[Int] = {
    nums.indices.foldLeft((Map[Int, Int](), Array[Int]())) { 
      case ((map, result), i) =>
        if (result.nonEmpty) {
          (map, result)
        } else {
          val complement = target - nums(i)
          if (map.contains(complement)) {
            (map, Array(map(complement), i))
          } else {
            (map + (nums(i) -> i), result)
          }
        }
    }._2
  }
}

File: csharp/LeetCode/ArraysAndHashing/TwoSum.cs
public class TwoSum
{
    public int[] Solution(int[] nums, int target)
    {
        var numMap = new Dictionary<int, int>();

        for (int i = 0; i < nums.Length; i++)
        {
            int complement = target - nums[i];

            if (numMap.ContainsKey(complement))
            {
                return new int[] {numMap[complement], i};
            }
            numMap[nums[i]] = i;
        }
        return new int[] {};
    }
}